types/src/proof/ensure_proof.rs

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module declares the macro ensure_proof!,
+//! which checks a condition for a proof and throws a ProofError
+//! if the proof fails.
+//! Currently, this macro only supports SparseMerkleProofs
+
+use crate::proof::definition::SparseMerkleProof;
+use crate::proof::proof_error::ProofError;
+use failure::Error;
+
+macro_rules! ensure_proof {
+    ($proof:ident, $cond:expr, $e:expr) => {
+        match $proof {
```

Comment:
if you have a match statement where you only want to act on a single type, it's generally cleaner to use 
`if let x = Type {}`

Reply:
I tried this, but clippy complains that this is an irrefutable pattern. It's doing this because clippy checks for all the call sites (in `types/src/proof/definition.rs`) and is seeing that `$proof` is always `SparseMerkleProof`  - which is true for now, but not guaranteed in the future.

From this, I see two options:
1. Make the same assumption as clippy and remove checking whether $proof is a `SparseMerkleProof` for now, and make changes for the future if/when other proofs call this macro. 
2. Keep this code, and add more match statements for other proof types that call this macro in the future.

Personally I think 2 is safer, but I'm open to other opinions!